crypto/threads_pthread.c

```diff
@@ -42,12 +44,490 @@
+#endif
+
+static CRYPTO_THREAD_LOCAL rcu_thr_key;
+
+struct rcu_thr_data;
+
+/*
+ * users is broken up into 2 parts
+ * bits 0-15 current readers
+ * bit 32-63 - ID
+ */
+# define READER_SHIFT 0
+# define ID_SHIFT 32
+# define READER_SIZE 16
+# define ID_SIZE 32
+
+# define READER_MASK     (((uint64_t)1 << READER_SIZE)-1)
+# define ID_MASK         (((uint64_t)1 << ID_SIZE)-1)
+# define READER_COUNT(x) (((uint64_t)(x) >> READER_SHIFT) & READER_MASK)
+# define ID_VAL(x)       (((uint64_t)(x) >> ID_SHIFT) & ID_MASK)
+# define VAL_READER      ((uint64_t)1 << READER_SHIFT)
+# define VAL_ID(x)       ((uint64_t)x << ID_SHIFT)
+
+/*
+ * This is the core of an rcu lock
+ * it tracks the the readers and writers
+ * for the current quiescence point for
+ * a given lock
+ * users is the 64 bit value that stores
+ * the READERS/ID as defined
+ * above
+ *
+ */
+struct rcu_qp {
+    uint64_t users;
+};
+
+struct thread_qp {
+    struct rcu_qp *qp;
+    struct rcu_lock_internal *lock;
+};
+
+#define MAX_QPS 10
+/*
+ * This is the per thread tracking data
+ * that is assigned to each thread participating
+ * in an rcu qp
+ *
+ * qp points to the qp that it last acquired
+ *
+ */
+struct rcu_thr_data {
+    struct thread_qp thread_qps[MAX_QPS];
+};
+
+/*
+ * This is the internal version of a CRYPTO_RCU_LOCK
+ * it is cast from CRYPTO_RCU_LOCK
+ */
+struct rcu_lock_internal {
+    CRYPTO_RCU_LOCK lock; /* Must be first */
+    struct rcu_cb_item *cb_items;
+    uint64_t id_ctr;
+    struct rcu_qp *qp_group;
+    size_t group_count;
+    uint64_t reader_idx;
+    uint32_t next_to_retire;
+    uint64_t current_alloc_idx;
+    uint32_t writers_alloced;
+    pthread_mutex_t write_lock;
+    pthread_mutex_t alloc_lock;
+    pthread_cond_t alloc_signal;
+    pthread_mutex_t prior_lock;
+    pthread_cond_t prior_signal;
+};
+
+/*
+ * Called on thread exit to free the pthread key
+ * associated with this thread, if any
+ */
+static void free_rcu_thr_data(void *ptr)
+{
+    struct rcu_thr_data *data = 
+                        (struct rcu_thr_data *)CRYPTO_THREAD_get_local(&rcu_thr_key);
+
+    if (data)
+        CRYPTO_free(data, __FILE__, __LINE__);
+    CRYPTO_THREAD_set_local(&rcu_thr_key, NULL);
+}
+
+static void ossl_rcu_init(void)
+{
+    CRYPTO_THREAD_init_local(&rcu_thr_key, NULL);
+}
+
+/* Read side acquisition of the current qp */
+static struct rcu_qp *get_hold_current_qp(CRYPTO_RCU_LOCK *lock)
+{
+    uint64_t qp_idx;
+    struct rcu_lock_internal *ilock =
+        (struct rcu_lock_internal *)lock;
+
+    /* get the current qp index */
+    for(;;) {
+        qp_idx = (uint64_t)__ATOMIC_LOAD_N(&ilock->reader_idx, __ATOMIC_ACQUIRE);
+        __ATOMIC_ADD_FETCH(&ilock->qp_group[qp_idx].users, VAL_READER,
+                           __ATOMIC_SEQ_CST);
+        /* if the idx hasn't changed, we're good, else try again */
+        if (qp_idx == (uint64_t)__ATOMIC_LOAD_N(&ilock->reader_idx, __ATOMIC_ACQUIRE))
```

Comment:
So the users is still SEQ_CST, and I have no idea why.

I think if you look at things properly, it should be possible to have the first (in this function) an acquire, the users increment a relaxed, and the check to see if it changed can go away. It requires that the order of the other atomic reads and write is correct, and that will need to have someone to write down things so it can be reasoned about. And we probably also need that for acquire/release.

Reply:
I don't see how the check for change can go away here.  We need to ensure that, once we have taken a read-side hold on a lock, that the write side hasn't updated the reader_idx.  Otherwise, its possible that the write side may race ahead of the reader, after we've read the reader_idx, but before we've incremented the user count, leading to a condition in which the read side is holding a data reference that the write side has already retired.  The result would be that the write side of a lock may free data before the read side is done reading it.

Reply:
I've documented the reasons for each memory constraint on the read and write side of the locks.